File: typescript/packages/core/src/resource/dify/dify.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import { DifyAccessor } from '../../accessor/dify.ts'
import { DIFY_COMMANDS } from '../../commands/builtin/dify/index.ts'
import type { RegisteredCommand } from '../../commands/config.ts'
import { makeResolveGlob } from '../../commands/builtin/generic_bind/index.ts'
import { readBytes } from '../../core/dify/read.ts'
import { readdir as difyReaddir } from '../../core/dify/readdir.ts'
import { stat as difyStat } from '../../core/dify/stat.ts'
import { DIFY_OPS } from '../../ops/dify/index.ts'
import type { RegisteredOp } from '../../ops/registry.ts'
import { ResourceName, type FileStat, type PathSpec } from '../../types.ts'
import { BaseResource, type Resource } from '../base.ts'
import { resolveDifyConfig, type DifyConfig, type DifyConfigResolved } from './config.ts'
import { DIFY_PROMPT } from './prompt.ts'

const resolveGlob = makeResolveGlob(difyReaddir)

export interface DifyResourceOptions {
  config: DifyConfig
}

export class DifyResource extends BaseResource implements Resource {
  readonly kind: string = ResourceName.DIFY
  readonly cachesReads: boolean = true
  readonly supportsSnapshot: boolean = false
  readonly prompt: string = DIFY_PROMPT
  readonly config: DifyConfigResolved
  readonly accessor: DifyAccessor

  constructor(options: DifyResourceOptions | DifyConfig) {
    super()
    const config = 'config' in options ? options.config : options
    this.config = resolveDifyConfig(config)
    this.accessor = new DifyAccessor(this.config)
  }

  open(): Promise<void> {
    return Promise.resolve()
  }

  close(): Promise<void> {
    return Promise.resolve()
  }

  ops(): readonly RegisteredOp[] {
    return DIFY_OPS
  }

  commands(): readonly RegisteredCommand[] {
    return DIFY_COMMANDS
  }

  glob(paths: readonly PathSpec[], _prefix = ''): Promise<PathSpec[]> {
    return resolveGlob(this.accessor, paths, this.index)
  }

  readFile(p: PathSpec): Promise<Uint8Array> {
    return readBytes(this.accessor, p, this.index)
  }

  readdir(p: PathSpec): Promise<string[]> {
    return difyReaddir(this.accessor, p, this.index)
  }

  stat(p: PathSpec): Promise<FileStat> {
    return difyStat(this.accessor, p, this.index)
  }
}

File: python/mirage/resource/dify/dify.py
from typing import Any

from mirage.accessor.dify import DifyAccessor
from mirage.commands.builtin.dify import COMMANDS
from mirage.core.dify.read import read_bytes, read_stream
from mirage.core.dify.readdir import readdir
from mirage.core.dify.stat import stat
from mirage.ops.dify import OPS as DIFY_VFS_OPS
from mirage.resource.base import BaseResource
from mirage.resource.dify.config import DifyConfig
from mirage.resource.dify.prompt import PROMPT
from mirage.types import ResourceName
from mirage.utils.glob_walk import make_resolve_glob

_resolve_glob = make_resolve_glob(readdir)

_DIFY_OPS = {
    "read_bytes": read_bytes,
    "read_stream": read_stream,
    "readdir": readdir,
    "stat": stat,
}


class DifyResource(BaseResource):

    accessor: DifyAccessor
    name: str = ResourceName.DIFY
    caches_reads: bool = True
    _ops = _DIFY_OPS
    PROMPT: str = PROMPT
    SUPPORTS_SNAPSHOT: bool = False

    def __init__(self, config: DifyConfig) -> None:
        super().__init__()
        self.config = config
        self.accessor = DifyAccessor(config)

        for fn in COMMANDS:
            self.register(fn)
        for fn in DIFY_VFS_OPS:
            self.register_op(fn)

    async def resolve_glob(self, paths, prefix: str = ""):
        return await _resolve_glob(self.accessor, paths, index=self._index)

    def get_state(self) -> dict[str, Any]:
        redacted = ["api_key"]
        config = self.config.model_dump()
        if config.get("api_key") is not None:
            config["api_key"] = "<REDACTED>"
        return {
            "type": self.name,
            "needs_override": True,
            "redacted_fields": redacted,
            "config": config,
        }

    def load_state(self, state: dict[str, Any]) -> None:
        pass
